qualtran/simulation/MBUC.ipynb

```diff
@@ -0,0 +1,375 @@
+{
+ "cells": [
+  {
+   "cell_type": "markdown",
+   "id": "44f40baf-3f87-40c5-9bf1-409ac3f86e68",
+   "metadata": {},
+   "source": [
+    "# Verifying Measurement-Based Uncomputation\n",
+    "\n",
+    "Quantum information cannot be destroyed, but during a computation we may produce intermediate values that we wish to discard. We can \"uncompute\" these values by running the computation in reverse. The ordinary uncomputation strategy requires paying the cost of the computation twice, but [*Halving the cost of quantum addition.* Gidney 2017](https://arxiv.org/abs/1709.06648) shows how measurement in the X basis can effectively discard a bit without expensive uncomputation. The consequence is that the remaining states of the system will pick up phases depending on the random measurement result. [*Verifying Measurement Based Uncomputation.* Gidney 2019](https://algassert.com/post/1903) provides more detail about these phases. It also describes a proceedure for using a phased-classical simulator to \"fuzz test\" measurement-based uncomputation circuits.\n",
+    "\n",
+    "Here, we show how Qualtran can be used to verify measurement based uncomputation circuits following Gidney's proposal."
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "3c3fdd7b-a94b-4f57-bb51-eec6217018df",
+   "metadata": {},
+   "source": [
+    "## Uncomputing $\\mathrm{And}$\n",
+    "\n",
+    "As a warm-up, we can use the reference classical action of `And(uncompute=True)` to verify the truth table of the operation. First, we check the bloq over valid inputs."
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "id": "3dddb0a5-a2b6-44cb-beb9-79b9600aae5e",
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "import itertools\n",
+    "from qualtran.bloqs.mcmt import And\n",
+    "\n",
+    "and_dag = And(uncompute=True)\n",
+    "for q1, q2 in itertools.product(range(2), repeat=2):\n",
+    "    trg = int(q1==1 and q2 == 1)\n",
```

Comment:
what is 'trg'?  I would probably use a better name.  
It took me a while to realize this means "target".  I would use that name here and below.

Avoid abbreviations:  https://google.github.io/styleguide/pyguide.html#316-naming